src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -249,7 +247,10 @@ public T Dequeue()
 
             T removed = _array[_head];
             _array[_head] = default(T);
-            _head = (_head + 1) % _array.Length;
+            if (++_head == _array.Length)
```

Comment:
:exclamation: In multithreaded code which does not correctly synchronize on usages of `Queue<T>`, this change introduces a point in time where a call to `Peek()` will throw an `ArrayIndexOutOfBoundsException`, where in the previous implementation it would return `default(T)` for the same interleaving. This is not a thread-safe type, but this change in observable behavior should be explicitly confirmed as desired.